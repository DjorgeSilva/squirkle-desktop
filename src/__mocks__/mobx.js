export const observable = require.requireActual('mobx').observable
export const autorun = jest.fn()
export const action = require.requireActual('mobx').action
export const spy = jest.fn() // needed for mobx-react-lite check
export const configure = jest.fn() // needed for mobx-react-lite check
export const toJS = jest.fn(x => x) // needed for mobx-react-lite check
