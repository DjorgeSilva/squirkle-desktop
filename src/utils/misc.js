// @flow

/**
 * Prepends a 0 if below 10.
 * @param {number} input The number.
 * @returns {string} The string with 0 if needed.
 */
function prepend0IfNeeded(input: number): string {
  return input < 10 ? `0${input}` : input.toString()
}

/**
 * Gets a Date as a nice format.
 * @param {Date} date The input date.
 * @returns {string} The formatted date.
 */
export function getFormattedDate(date: Date): string {
  return `${date.getHours()}:${prepend0IfNeeded(
    date.getMinutes()
  )}:${prepend0IfNeeded(date.getSeconds())}:${date.getMilliseconds()}`
}

/**
 * Checks if in Electron or not.
 * @returns {boolean} If in Electron.
 */
export function isElectronApp(): boolean {
  return navigator.userAgent.includes('Electron')
}