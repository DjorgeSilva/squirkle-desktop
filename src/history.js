// @flow

import { createHashHistory, createBrowserHistory } from 'history'
import { isElectronApp } from './utils/misc'

export default isElectronApp() ? createHashHistory() : createBrowserHistory()
