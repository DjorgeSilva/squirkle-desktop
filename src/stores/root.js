// @flow

import { observable } from 'mobx'
import { isElectronApp } from '../utils/misc'
import { syncStore } from '../utils/stores'

export type RootStoreType = {
  version: string,
  tasks: Array<any>,
  loadAllTasks: () => Array<any>
}


export default () => {
  const store = observable({
    version: '1.0',
    tasks: [],

    loadAllTasks (): Array<any> {
      console.log('View called loadAllTasks')
      return []
    }
  },
  // Mobx - Effects
  {}
  )

  if (isElectronApp()) {
    syncStore('root', store)
  }

  return store
}