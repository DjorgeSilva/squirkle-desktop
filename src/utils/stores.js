// @flow

import { syncStore as syncDesktopStore } from '@atomos/atomizer-desktop/client/stores'

/**
* Syncs the store with other desktop windows.
* @param {string} storeName The name of the store.
* @param {store} store The store.
* @return {void}
*/
export function syncStore(storeName: string, store: any): void {
  syncDesktopStore(storeName, store)
}