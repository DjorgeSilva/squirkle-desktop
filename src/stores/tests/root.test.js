// @flow

import { destroy as destroyWindow } from '@atomos/atomizer-desktop/client/window'
import { syncStore } from '@atomos/atomizer-desktop/client/stores'
import { autorun } from 'mobx'
import createRootStore from '../root'
import history from '../../history'
import { isElectronApp } from '../../utils/misc'
import { changeLanguage } from '@atomos/atomizer-utils/i18n'

jest.mock('mobx')
jest.mock('@atomos/atomizer-desktop/client/window')
jest.mock('@atomos/atomizer-desktop/client/stores')
jest.mock('@atomos/atomizer-utils/i18n')

jest.mock('../../history', () => ({
  __esModule: true,
  default: {
    push: jest.fn()
  }
}))
jest.mock('../../utils', () => ({
  isElectronApp: jest.fn()
}))

describe('Root store', (): void => {
  describe('When created', (): void => {
    let store

    beforeAll((): void => {
      ;(isElectronApp: any).mockReturnValue(true)
      store = createRootStore()
    })

    it('initializes default properties', (): void => {
      expect(store).toMatchSnapshot()
    })

    it('tells mobx to autorun', (): void => {
      expect(autorun).toHaveBeenCalled()
    })

    it('syncs the store', (): void => {
      expect(syncStore).toHaveBeenCalledWith('Root', store)
    })

    describe('then the autorun is called', (): void => {
      beforeAll((): void => {
        const autorunSideEffect: any = autorun.mock.calls[0][0]
        autorunSideEffect()
      })

      it('changes the i18n language', (): void => {
        expect(changeLanguage).toHaveBeenCalledWith('en')
      })
    })

    describe('then we login', (): void => {
      beforeAll((): void => {
        store.login('Joe Bloggs')
      })

      it('sets authenticated', (): void => {
        expect(store.isAuthenticated).toBe(true)
      })

      it('sets username', (): void => {
        expect(store.userName).toBe('Joe Bloggs')
      })

      it('destroys the login window', (): void => {
        expect(destroyWindow).toHaveBeenCalledWith('login')
      })

      it('navigates to home', (): void => {
        expect(history.push).toHaveBeenCalledWith('/')
      })
    })

    describe('then we set an album', (): void => {
      beforeAll((): void => {
        store.setAlbum(123, {
          id: 123,
          name: 'Discovery'
        })
      })

      it('updates the store', (): void => {
        expect(store.albums).toEqual([{ id: 123, name: 'Discovery' }])
      })
    })

    describe('then we set an artist', (): void => {
      beforeAll((): void => {
        store.setArtist(456, {
          id: 456,
          name: 'Daft Punk'
        })
      })

      it('updates the store', (): void => {
        expect(store.artists).toEqual([{ id: 456, name: 'Daft Punk' }])
      })
    })

    describe('then we add songs', (): void => {
      beforeAll((): void => {
        store.addSongs([
          {
            id: 111,
            title: 'One More Time'
          }
        ])
      })

      it('updates the store', (): void => {
        expect(store.songs).toEqual([{ id: 111, title: 'One More Time' }])
      })
    })

    describe('then we add a song ID to a playlist', (): void => {
      beforeAll((): void => {
        store.addSongIdToPlaylist(1, 111)
      })

      it('updates the store', (): void => {
        expect(store.playlists).toEqual([
          { id: 1, name: 'Liked Songs', songIds: [32861727, 111] }
        ])
      })

      describe('then we remove it', (): void => {
        beforeAll((): void => {
          store.removeSongIdFromPlaylist(1, 111)
        })

        it('updates the store', (): void => {
          expect(store.playlists).toEqual([
            { id: 1, name: 'Liked Songs', songIds: [32861727] }
          ])
        })
      })
    })

    describe('then we set song IDs to play', (): void => {
      beforeAll((): void => {
        store.setSongIdsToPlay([111, 222, 333])
      })

      it('updates the store', (): void => {
        expect(store.songIdsToPlay).toEqual([111, 222, 333])
        expect(store.nowPlayingSongId).toBe(111)
      })
    })
  })
})
