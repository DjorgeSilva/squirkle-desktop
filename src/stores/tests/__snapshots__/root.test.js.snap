// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root store When created initializes default properties 1`] = `
Object {
  "activeLanguage": "en",
  "addArtists": [Function],
  "addSongIdToPlaylist": [Function],
  "addSongs": [Function],
  "albums": Array [],
  "artists": Array [],
  "avatarUrl": null,
  "currentRouteName": null,
  "currentSongDuration": 0,
  "currentTime": 0,
  "isAuthenticated": false,
  "isPlaying": false,
  "login": [Function],
  "logout": [Function],
  "nowPlayingSongId": null,
  "playLocalFileUrl": "",
  "playlists": Array [
    Object {
      "id": 1,
      "name": "Liked Songs",
      "songIds": Array [
        32861727,
      ],
    },
  ],
  "recommendedAlbumIds": Array [],
  "removeSongIdFromPlaylist": [Function],
  "seekTime": 0,
  "setAlbums": [Function],
  "setArtist": [Function],
  "setArtists": [Function],
  "setAvatarUrl": [Function],
  "setCurrentRouteName": [Function],
  "setCurrentTime": [Function],
  "setIsPlaying": [Function],
  "setPlayLocalFileUrl": [Function],
  "setPlaySongById": [Function],
  "setPlaylists": [Function],
  "setRecommendedAlbumIds": [Function],
  "setSeekTime": [Function],
  "setSongDuration": [Function],
  "setSongIdsToPlay": [Function],
  "setSongs": [Function],
  "setUserName": [Function],
  "setVolumePercent": [Function],
  "songIdsToPlay": Array [],
  "songs": Array [],
  "userName": "",
  "volumePercent": 40,
}
`;
